### PR TITLE
Add ability to logout

### PR DESCRIPTION
**Pull request by Jamiltz**
See https://github.com/couchbaselabs/ToDoLite-Android/issues/8

### CODE REVIEW

**Comment by tleyden** [comment]
@Jamiltz it's not mergable .. were your changes based off of current master or an old version of master?  Maybe you need to rebase?

**Comment by Jamiltz** [comment]
Yes it was behind master. Rebased.

File: ToDoLite/src/main/java/com/couchbase/todolite/Application.java [hunk 1]
```diff
@@ -34,8 +37,8 @@
 
     public static final String TAG = "ToDoLite";
     private static final String DATABASE_NAME = "todos";
-    private static final String SYNC_URL_HTTP = "http://demo-mobile.couchbase.com/todolite/";
-    private static final String SYNC_URL_HTTPS = "https://demo-mobile.couchbase.com/todolite/";
+    private static final String SYNC_URL_HTTP = "http://demo.mobile.couchbase.com/todolite/";
+    private static final String SYNC_URL_HTTPS = "https://demo.mobile.couchbase.com/todolite/";
```
Comment (tleyden) [inline ToDoLite/src/main/java/com/couchbase/todolite/Application.java]: for https, only demo-mobile.couchbase.com works.

**Comment by Jamiltz** [comment]
Oops, selected the wrong side in difftool. Updated domains to demo-mobile.

File: ToDoLite/src/main/java/com/couchbase/todolite/Application.java [hunk 2]
```diff
@@ -298,11 +210,34 @@ public void onCreate() {
...
+        MessageDigest digest = null;
+        try {
+            digest = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
+        }
+        try {
+            byte[] inputBytes = name.getBytes();
+            byte[] hashBytes = digest.digest(inputBytes);
+            database = manager.getDatabase(byteArrayToHex(hashBytes));
```
Comment (pasin) [inline ToDoLite/src/main/java/com/couchbase/todolite/Application.java]: IIRC, db name cannot begin with number so to make sure that the db name is valid, you will need to prefix the name with some characters such as 'db'.

File: ToDoLite/src/main/java/com/couchbase/todolite/LoginActivity.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+        mCookieAuthButton = (Button)findViewById(R.id.cookieAuthLoginButton);
+        mCookieAuthButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                loginWithCustomCookieAndStartSync();
+            }
+        });
+
+    }
```
Comment (pasin) [inline ToDoLite/src/main/java/com/couchbase/todolite/LoginActivity.java]: Can you also add guest login? We kind of have this feature in the current app.

**Comment by pasin** [comment]
Apart from the inline comment above,

1. The code doesn't perform migration from the old database structure as we are moving toward to one user, one database. I think this is OK , and users should get their documents synced from the sync gateway.

2. I don't know if we should keep login with basic and with custom cookies. This might be good for debugging various types of authentication but they are not useful for users. Should we hide them by default and provide a flag to enable them in code? @tleyden , what is your opinion on this?

**Comment by tleyden** [comment]
> Should we hide them by default and provide a flag to enable them in code?

Yes, I agree with that approach.  If we upgrade the demo cluster to have a better "custom auth" story, then we can build in custom cookie based auth into the app and give users the choice.  Until then, it should be disabled by default.

**Comment by Jamiltz** [comment]
Guest account should be working and buttons for cookie and custom auth are hidden by default.
It should behave like http://f.cl.ly/items/0W2H3S3p2P3d10001k0m/Guest%20Account.mp4

File: ToDoLite/src/main/java/com/couchbase/todolite/Application.java [hunk 4]
```diff
@@ -95,148 +112,147 @@ private synchronized void updateSyncProgress(int completedCount, int totalCount,
...
-    public void startReplicationSyncWithBasicAuth(String username, String password) {
+            QueryEnumerator rows = null;
+            try {
+                rows = guestDb.createAllDocumentsQuery().run();
+            } catch (CouchbaseLiteException e) {
+                Log.e(TAG, "Cannot run all docs query", e);
+            }
 
-        if (pullReplication == null && pushReplication == null) {
+            if (rows.getCount() == 0) {
```
Comment (pasin) [inline ToDoLite/src/main/java/com/couchbase/todolite/Application.java]: Nothing wrong with this but the code might be cleaner with just calling guestDb.getDocumentCount().

**Pull request merged**